# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [1]:
import os
import numpy as np
import pandas as pd

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [5]:
print([numbers for numbers in range(50)])

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]


### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [6]:
print([evens for evens in range(50) if evens %2 == 0])

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48]


### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [16]:
a = np.array([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
       [0.44409793, 0.09014516, 0.81835917, 0.87645456],
       [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
       [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
       [0.17003563, 0.46843998, 0.92796258, 0.69814654],
       [0.41290051, 0.19561071, 0.16284783, 0.97016248],
       [0.71725408, 0.87702738, 0.31244595, 0.76615487],
       [0.20754036, 0.57871812, 0.07214068, 0.40356048],
       [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
       [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

In [36]:
print(a)

[[0.84062117 0.48006452 0.7876326  0.77109654]
 [0.44409793 0.09014516 0.81835917 0.87645456]
 [0.7066597  0.09610873 0.41247947 0.57433389]
 [0.29960807 0.42315023 0.34452557 0.4751035 ]
 [0.17003563 0.46843998 0.92796258 0.69814654]
 [0.41290051 0.19561071 0.16284783 0.97016248]
 [0.71725408 0.87702738 0.31244595 0.76615487]
 [0.20754036 0.57871812 0.07214068 0.40356048]
 [0.12149553 0.53222417 0.9976855  0.12536346]
 [0.80930099 0.50962849 0.94555126 0.33364763]]


In [39]:
numpy_to_list = [y for e in a for y in e]
print(numpy_to_list)

[0.84062117, 0.48006452, 0.7876326, 0.77109654, 0.44409793, 0.09014516, 0.81835917, 0.87645456, 0.7066597, 0.09610873, 0.41247947, 0.57433389, 0.29960807, 0.42315023, 0.34452557, 0.4751035, 0.17003563, 0.46843998, 0.92796258, 0.69814654, 0.41290051, 0.19561071, 0.16284783, 0.97016248, 0.71725408, 0.87702738, 0.31244595, 0.76615487, 0.20754036, 0.57871812, 0.07214068, 0.40356048, 0.12149553, 0.53222417, 0.9976855, 0.12536346, 0.80930099, 0.50962849, 0.94555126, 0.33364763]


### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [56]:
numpy_to_list = [y for e in a for y in e if y >= 0.5]
print(numpy_to_list)

[0.84062117, 0.7876326, 0.77109654, 0.81835917, 0.87645456, 0.7066597, 0.57433389, 0.92796258, 0.69814654, 0.97016248, 0.71725408, 0.87702738, 0.76615487, 0.57871812, 0.53222417, 0.9976855, 0.80930099, 0.50962849, 0.94555126]


### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [59]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])

In [65]:
desanidar_3 = [des3 for des1 in b for des2 in des1 for des3 in des2]
print(desanidar_3)

[0.55867166, 0.06210792, 0.08147297, 0.82579068, 0.91512478, 0.06833034, 0.05440634, 0.65857693, 0.30296619, 0.06769833, 0.96031863, 0.51293743, 0.09143215, 0.71893382, 0.45850679, 0.58256464, 0.59005654, 0.56266457, 0.71600294, 0.87392666, 0.11434044, 0.8694668, 0.65669313, 0.10708681, 0.07529684, 0.46470767, 0.47984544, 0.65368638, 0.14901286, 0.23760688]


### 5. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [126]:
desanidar_4 = [des2[2] for des1 in b for des2 in des1 if des2[2] <= 0.5]
print(desanidar_4)

[0.08147297, 0.06833034, 0.30296619, 0.45850679, 0.11434044, 0.10708681, 0.47984544, 0.23760688]


### 6. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

In [130]:

[filelist for filelist in os.listdir("../data") if ".csv" in filelist]

['sample_file_0.csv',
 'sample_file_1.csv',
 'sample_file_2.csv',
 'sample_file_3.csv',
 'sample_file_4.csv',
 'sample_file_5.csv',
 'sample_file_6.csv',
 'sample_file_7.csv',
 'sample_file_8.csv',
 'sample_file_9.csv']

### 7. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

In [141]:
file = pd.read_csv("../data/sample_file_0.csv")
pd.set_option("display.max_row", 10)
print(file)


          0         1         2         3         4         5         6  \
0  0.734751  0.195362  0.734309  0.598184  0.763433  0.263434  0.868066   
1  0.772607  0.445391  0.249642  0.787922  0.598583  0.827238  0.624126   
2  0.226428  0.268764  0.694262  0.622335  0.063843  0.122683  0.815625   
3  0.362748  0.495430  0.113876  0.594149  0.612522  0.625204  0.864050   
4  0.033415  0.340433  0.464971  0.363737  0.025815  0.434129  0.415163   

          7         8         9        10        11        12        13  \
0  0.058092  0.753502  0.587513  0.311608  0.178356  0.182922  0.147631   
1  0.601524  0.688753  0.338870  0.081595  0.471474  0.267443  0.453351   
2  0.584542  0.032594  0.589775  0.764350  0.650973  0.565705  0.691784   
3  0.260279  0.528873  0.168043  0.715929  0.677014  0.175735  0.632370   
4  0.892210  0.381701  0.415264  0.790801  0.696930  0.819751  0.944029   

         14        15        16        17        18        19  
0  0.391188  0.816049  0.749068  0

### 8. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 9. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 10. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.